node/config/types.go

```diff
@@ -167,6 +167,16 @@ type DealmakingConfig struct {
 
 	// The maximum amount of time a transfer can take before it fails
 	MaxTransferDuration Duration
+
+	// The maximum number of concurrent storage deal HTTP downloads.
+	// Note that this is a soft maximum; if some downloads stall,
+	// more downloads are allowed to start.
+	HttpTransferMaxConcurrentDownloads uint64
```

Comment:
This 'soft' limit probably only makes sense if we're not starting more streams with a peer which has stalled connections - e.g. when we're retrieving 20 streams at 1mb/s (where it's likely that some streams will stall for >30s) from a very overloaded estuary shuttle, starting more streams will only make things worse.

Reply:
Having hard per-peer limit could potentially help here

Reply:
FWIW on our systems we want very high per host concurrency as they are general purpose HTTP servers and so routing is tuned for fairness and each connection is quite slow but overall throughput is very high and scales linearly with stream count (this is fairly common with many HTTP hosts, thus the popularity of "download accelerators"). To some extent this also happens naturally if TCP congestion control get stuck in avoidance mode.

Ideally not only would we have multiple parallel file downloads per host but also multipart within each file, like `aria2c --split=X --max-connection-per-server=X*n...` with X > 16. I believe some tooling like Filswan already does this with aria2.

Reply:
To expand on the above a bit: obviously the two scenarios we are imagining are nearly polar opposites, and difficult to tell apart by heuristics on the boostd side. One immediate solution would be to make this configurable by the client (`--max-connection-per-peer`), with a default representing whichever case you think is most likely for a naive user.

For reference, we're currently doing 150 concurrent transfers per SP comfortably and testing how much higher it can go, this could easily go into thousands with multipart (scaling up our transfer capacity accordingly)

Reply:
@parkan thanks for the comments, that makes sense 👍 
I've added some context to the PR description to make clear why SPs want this change.
For your use case I think the SP will still want to have some limit on the number of concurrent transfers, to make sure their sealing pipeline is always busy. It will just be a higher number. But certainly we will have an option whereby SPs can disable transfer limiting entirely.

Reply:
@magik6k thanks for pointing that out. I agree that we should not open more connections to a peer that already has stalled transfers. I'll make that change 👍 

I would prefer to try to keep the number of configuration knobs to a minimum because in the past transfer limits have caused a lot of confusion for SPs. For example the graphsync download limit would often result in the download pipeline getting completely blocked up and SPs wouldn't know why they weren't getting any data.

To improve visibility we're going to show the transfer rate for each deal on the deals page and also in summary format on the transfers page (as well as https://github.com/filecoin-project/boost/issues/676).

I'd suggest we try out this simple change first and see how it works, then add more knobs if it's not sophisticated enough.

Reply:
@dirkmc totally understand the SP side of this (one of mine actually ran into a stampede situation recently as well), just want to make sure we aren't getting shut out in the process. If the limits are high enough it may have limited impact on us as long as we get enough SPs.